src/ast/passes/semantic_analyser.cpp

```diff
@@ -2040,7 +2044,7 @@ void SemanticAnalyser::visit(ArrayAccess &arr)
         auto index = bpftrace_.get_int_literal(arr.indexpr);
         if (index.has_value()) {
           size_t num = type.GetNumElements();
-          if (num != 0 && static_cast<size_t>(*index) >= num)
+          if (num != 0 && std::cmp_greater_equal(*index, num))
```

Comment:
This one is causing the issue with the `semantic_analyser.array_access` test: when `*index` is negative, the comparison always returns `false`, since `num` is unsigned.